data/src/main/java/org/apache/iceberg/data/BaseFileWriterFactory.java

```diff
@@ -62,7 +63,11 @@ protected BaseFileWriterFactory(
       Schema positionDeleteRowSchema) {
     this.table = table;
     this.dataFileFormat = dataFileFormat;
-    this.dataSchema = dataSchema;
+    if (dataSchema != null) {
+      this.dataSchema = PruneUnknownTypes.convert(dataSchema);
```

Comment:
What if we just made `convert` return `null` for a null input? Then this `if` can be in that method rather than in all the places where it is called.